searchlib/src/vespa/searchlib/tensor/hnsw_index_loader.cpp

```diff
@@ -0,0 +1,48 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include "hnsw_index_loader.h"
+#include "hnsw_graph.h"
+#include <vespa/searchlib/util/fileutil.h>
+
+namespace search::tensor {
+
+HnswIndexLoader::~HnswIndexLoader() {}
+
+HnswIndexLoader::HnswIndexLoader(HnswGraph &graph)
+    : _graph(graph)
+{
+}
+
+bool
+HnswIndexLoader::load(const fileutil::LoadedBuffer& buf)
+{
+    size_t num_readable = buf.size(sizeof(uint32_t));
+    if (num_readable < 3) return false;
+    const uint32_t *ptr = static_cast<const uint32_t *>(buf.buffer());
+    const uint32_t *end = ptr + num_readable;
+    uint32_t entry_docid = *ptr++;
+    int32_t entry_level = *ptr++;
+    uint32_t num_nodes = *ptr++;
+    std::vector<uint32_t> link_array;
+    for (uint32_t docid = 0; docid < num_nodes; ++docid) {
+        if (ptr == end) return false;
+        uint32_t num_levels = *ptr++;
+        if (num_levels > 0) {
```

Comment:
If num_levels is 0 (representing a removed document), we should create an entry for this document to ensure that node_refs has the right size. If the last document loaded has 0 levels (removed) the size of node_refs will be off by one.